Sources/Foundation/JSONEncoder.swift

```diff
@@ -237,176 +237,405 @@ open class JSONEncoder {
     /// - returns: A new `Data` value containing the encoded JSON data.
     /// - throws: `EncodingError.invalidValue` if a non-conforming floating-point value is encountered during encoding, and the encoding strategy is `.throw`.
     /// - throws: An error if any value throws an error during encoding.
-    open func encode<T : Encodable>(_ value: T) throws -> Data {
-        let encoder = _JSONEncoder(options: self.options)
- 
-        guard let topLevel = try encoder.box_(value) else {
+    open func encode<T: Encodable>(_ value: T) throws -> Data {
+        let value: JSONValue = try encodeAsJSONValue(value)
+        let writer = JSONValue.Writer(options: self.outputFormatting)
+        let bytes = writer.writeValue(value)
+        
+        return Data(bytes)
+    }
+
+    func encodeAsJSONValue<T: Encodable>(_ value: T) throws -> JSONValue {
+        let encoder = JSONEncoderImpl(options: self.options, codingPath: [])
+        guard let topLevel = try encoder.wrapEncodable(value, for: nil) else {
             throw EncodingError.invalidValue(value, EncodingError.Context(codingPath: [], debugDescription: "Top-level \(T.self) did not encode any values."))
         }
+        
+        return topLevel
+    }
+}
 
-        let writingOptions = JSONSerialization.WritingOptions(rawValue: self.outputFormatting.rawValue).union(.fragmentsAllowed)
+// MARK: - _JSONEncoder
 
-        do {
-            return try JSONSerialization.data(withJSONObject: topLevel, options: writingOptions)
-        } catch {
-            throw EncodingError.invalidValue(value, EncodingError.Context(codingPath: [], debugDescription: "Unable to encode the given top-level value to JSON.", underlyingError: error))
+private enum JSONFuture {
+    case value(JSONValue)
+    case encoder(JSONEncoderImpl)
+    case nestedArray(Array)
+    case nestedObject(Object)
+    
+    class Array {
```

Comment:
Could we name this something more specific? I get it's a private type but that `case nestedArray(Array)` threw me for a second.